### ISSUE TITLE
Please add the posibility to configure the router ip that is advertise

### ISSUE THREAD

**Issue by ruben-herold** [reubenhwk/radvd#45]
hi,

would it possible to add support for configuring a ipv6 ip which is advertised as router by the radvd service? This would it make possible to run radvd on system in the network if you have an device as gateway which don't support router advertisment. 

Also if you run vrrp like keepalived for router redundancy on your connection tracking firewalls, it would be very helpfull only advertise only the service ip to avoid asyncron routing which brakes connection tracking.

**Comment by reubenhwk** [comment]
Hi Ruben,

I'm having a difficult time understanding your question, but I'll try to
answer the question I think you're asking...

Yes, any system on a network can run RADVD and advertise the IPv6 network
prefix.  That system does not need to be the router or gateway.

I don't know how RADVD is related to VRRP and connection tracking.  Can you
elaborate?

**Comment by ruben-herold** [comment]
hi,

I have attached a graph about the setup:
![ravd](https://cloud.githubusercontent.com/assets/1569572/7684533/15f7e8c6-fd87-11e4-828f-053034a1ff64.png)

As you can see the IP 2001:DB8::1/64 is moved via VRRP between the two firewalls. 
If I install radvd on both firewalls, the anounce both their link local adresses as gateway 
and I have an async. routing so that connection tracking will fail.

From my point of view there are only this solution:
 
a) start radvd only on the firewall which has the IP Adresss activ and try to automate the start and   the stop 

b) start radvd on both and have the possibilty to anounce 2001:DB8::1 as gateway from both.

**Comment by ruben-herold** [comment]
If take a look around how other platforms handle this issue:

Junos:

[edit protocols router-advertisement interface ge-0/0/1.0]
router# set ?
Possible completions:
  [...]
  virtual-router-only  Send advertisemnets only for vrrp-inet6-group
[edit protocols router-advertisement interface ge-0/0/1.0]

**Comment by reubenhwk** [comment]
Is there an RFC which defines this feature?

**Comment by ruben-herold** [comment]
I don't think. There are RFC for VRRP (RFC 5798) but not for this special situation. Most vendors have integrated it like juniper.  You can also take a look at:

https://www.ietf.org/mail-archive/web/vrrp/current/msg00764.html

The linux VRRP Implementation (keepalived) can't do router advertisment, so the users must use keepalived together with radvd to get router advertisments.

**Comment by ruben-herold** [comment]
Do you need any further informations?

**Comment by reubenhwk** [comment]
No.  Unfortunately I don't have time to do this.  I'll be happy to commit
source code if you can fork radvd and make the changes.

**Comment by ruben-herold** [comment]
I will try to read through the source, but my C knowledge is very low. But if I see it the right way:

a) config option for each prefix which router should be announced
b) if config option is set send ra packet with this src ip6

**Comment by reubenhwk** [comment]
On Wed, Jun 3, 2015 at 2:08 AM, Ruben Herold <notifications@github.com>
wrote:

> I will try to read through the source, but my C knowledge is very low. But
> if I see it the right way:
>
> a) config option for each prefix which router should be announced
>
See in gram.y 'prefixparms'.  I think what you'll need to do is add a
section under prefixparms...something like this...

| T_MyNewOption IPV6ADDR
{
    ...new code...
}

You'll need to add a '%token T_MyNewOption', then you'll have to add your
option to scanner.l.  Something like this...

MyNewOption { return T_MyNewOption; }

The structure you'll need to update is 'struct AdvPrefix' in radvd.h.  You
maybe will need to add another 'struct in6_addr' so you'll have a place to
store the IPV6ADDR from the code you add in gram.y.

b) if config option is set send ra packet with this src ip6

**Comment by reubenhwk** [comment]
Also, there's an option to run radvd and only check the syntax of the
config file.  Use the -C and -c flags to test your code.

**Comment by 0xbabaf00l** [comment]
Hi,

any progress on that issue?

We ran into the same problem.
Currently, I added a script to start radvd when VRRP is in master state and stop radvd when in BACKUP or FAIL. So I to have radvd running only on the active firewall.
This is not perfect, because the client has to learn its new default router on failover and so it does take several seconds.

**Comment by reubenhwk** [comment]
I'm looking into this, but I don't really understand what to do.  Can one of you explain what RADVD is doing now and what RADVD needs to do?

My initial thought is to do something like the following (but I don't know if this is right)...

Add a config option to override the link-local addr of an iface and use that addr in sendmsg.  Here's a snit of the diff...

-       memcpy(&pkt_info->ipi6_addr, &props->if_addr, sizeof(struct in6_addr));
+       if (props->if_addr_cfg) {
+               memcpy(&pkt_info->ipi6_addr, props->if_addr_cfg, sizeof(struct in6_addr));
+       } else {
+               memcpy(&pkt_info->ipi6_addr, &props->if_addr, sizeof(struct in6_addr));
+       }

This doesn't work.  sendmsg fails.  I may need to investigate using a different type of socket to make this work (RAW/PACKET/Whatever).

**Comment by reubenhwk** [comment]
So spoof the router IP?

**Comment by robbat2** [comment]
Rather than spoof the router IP, radvd could dynamically check if the specified address is present on the interface, and only then use it. This would tie in well with code for the getifaddrs `::/64` path and addresses being dynamically added to an interface.

I'd use VRRP with `fe80::1/64` address, having it on only one router at a time. I'd want to make sure `AdvSourceLLAddress` was set to `off` as well.

**Comment by reubenhwk** [comment]
Makes sense!  Also much simpler than what I was thinking.  I'll take a stab
at this when I get some free time.  Is there going to be anybody available
in the next couple week who can test it out to verify the changes?

On Wed, Apr 27, 2016 at 11:08 AM, Robin H. Johnson <notifications@github.com

**Comment by robbat2** [comment]
I could certainly test it.

I do see alternative use-cases for spoofing the source address, but that's entirely separate from this problem, and I'd want it handled separately (specifically, working around dumb routers that aren't as capable as radvd for RAs, but still letting them do the actual routing).

**Comment by ruben-herold** [comment]
sorry for my late answer, can someone tell me the current state?

**Comment by reubenhwk** [comment]
I recall looking at it, having trouble, and giving up.  There's a
possibility I didn't completely understand the problem, or the solution.  I
was under the impression I would have to spoof src addrs in the packet, but
I wasn't able to setup a raw socket to do that (not that it couldn't be
done, I just wasn't able to at the time).

**Comment by robbat2** [comment]
I did read more of the source, but never came back to implement.
```
send.c:774    struct in6_pktinfo *pkt_info = (struct in6_pktinfo *)CMSG_DATA(cmsg);
send.c:775    pkt_info->ipi6_ifindex = props->if_index;
send.c:776    memcpy(&pkt_info->ipi6_addr, &props->if_addr, sizeof(struct in6_addr));
```

Instead of the first link-local address, implement something that checks the array  `props->if_addrs` for a specified value (eg `fe80::1`). If present, use it; If not present, do NOT send the packet.

**Comment by robbat2** [comment]
Additionally here, the if you use the user-specified address blindly, without checking `props->if_addrs`, you should get the spoofing behavior trivially (I think the kernel will swallow it if you're non-root).

**Comment by reubenhwk** [comment]
What part of the code are you taking about Robin?  If I hack around with it
in send.c, I can do two things.  I can comment out the memcpy, in which
case I still get adverts.  I can also put something else in
pkt_info->ipi6_addr, in which case I don't get adverts.

It seems like this feature request is to put something other than the link
local address of the interface into that field, but so far, I can't.

**Comment by robbat2** [comment]
An interface may have MORE than one link-local address. In my case, I'd like to have `fe80::1` present on the interface and used in the advert.

**Comment by reubenhwk** [comment]
Ah!  I think I'm starting to get it...

So RADVD only reads (I think) the first link local addr it finds, and uses
that one.  You want a setting in the config file to specify which link
local addr to use when sending the advert?

If I recall correctly, the reason for this is so the hosts in the subnet
will configure their default route to use the link local addr from which
the router RA came, and you want to run RADVD on a device that is *not* the
default gateway.  Is that all correct?

I'm not sure, but I don't think the if_addr address matters.  I'm under the
impression that value is only to help the kernel know on which interface to
send the RA using a raw socket.  I think the setting you're interested in
is sllao->if_hwaddr.  The setting, AdvSourceLLAddress, is an on/off
setting, but maybe it can be extended to be the hardware address of the
default gateway.  I admin that I may be completely wrong.

Something like this may be exactly what radvd needs to do...

$ cat radvd.conf

eth0 {
  AdvSourceLLAddress 00:11:22:33:44:55;
};

$

...however I think it's a bad idea to assume a mac address (who knows what
type of interface hardware address it could actually be)...but that doesn't
mean we can't try it!

If this all makes sense to you, let me know.  I'll try to get this coded up
by next weekend and put it in a branch for you (don't hesitate to remind me
with obnoxious emails because I tend to get distracted by my day job).

-Reuben

**Comment by robbat2** [comment]
Running RADVD on a different host than the gateway is a possible edge case, but the most trivial one is just having multiple LLv6 addresses on the gateway, and allowing selection of the one you want. (eg a single `fe80::1` address in the network, and it moves around to the active gateway)

**Comment by greyslater** [comment]
I didn't fully understand
Can I use this to realy "spoof" the address of my acutal router?

My setup:

I have a Juniper SRX100 (Router) which can do RA but **not RRDNS**
so my idea was obvious.

use radvd on another box to do a RA including also RRDNS
but here the problem starts because radvd presents it's own LLv6 address as router instead of my Junipers address

another solution for me would to "split" the RA between my Router and radvd
- radvd could only do the RRDNS but I guess this is absolutely not supported

Thanks for your answers

**Comment by greyslater** [comment]
I found a solution myself:

My Juniper SRX100 (Router) is configured to do the RA
but because with JunOS 12.1 (which is maximum for SRX100) it can't do RRDNS (which was implemented not before JunOS 14.1)

another box with radvd is doing the RRDNS without "disturbing" my routing part

This is achived by 

`
interface eth0 {AdvSendAdvert on;
    AdvDefaultLifetime 0;
    prefix 2001:470:1f1b:15f::/64 {
        AdvOnLink on;
        AdvAutonomous on;
    };
    RDNSS 2001:470:1f1b:15f::2:10 {
        AdvRDNSSLifetime 1800;
    };
};
`

**Comment by greyslater** [comment]
I found a solution myself:

My Juniper SRX100 (Router) is configured to do the RA
but because with JunOS 12.1 (which is maximum for SRX100) **it can't do RRDNS** (which was implemented not before JunOS 14.1)

another box with **radvd is doing the RRDNS** without "disturbing" my routing part

This is achived by **AdvDefaultLifetime 0;** the clients won't use it as route and take the RRDNS portion only

interface eth0 {
    AdvSendAdvert on;
    AdvDefaultLifetime 0;
    prefix 2001:470:<my-net>::/64 {
        AdvOnLink on;
        AdvAutonomous on;
    };
    RDNSS 2001:470:<my-net>::1 {
        AdvRDNSSLifetime 1800;
    };
};

**Comment by candlerb** [comment]
For your RDNSS issue, another option would be to use `AdvOtherConfigFlag on` plus a stateless DHCPv6 server to hand out the DNS settings. The clients I've tried are pretty good about merging info from RAs and DHCP.
